stablehlo/dialect/StablehloOps.td

```diff
@@ -2491,11 +2433,7 @@ def StableHLO_ReduceWindowOp: StableHLO_Op<"reduce_window", [
     // Relax the strict default implementation with one that allows
     // for StableHLO-specific differences.
     static bool isCompatibleReturnTypes(TypeRange l, TypeRange r) {
-      if (l.size() != r.size()) return false;
-      for (auto [lt, rt] : llvm::zip(l, r))
-        if (!mlir::hlo::isCompatibleForHloTypeInference(lt, rt))
-          return false;
-      return true;
+      return mlir::hlo::isCompatibleForHloTypeInference(l, r);
     }
   }];
```

Comment:
Why do we still need this extraClassDeclaration here, if it's already defined in StableHLO_Op? Same question for StableHLO_UnaryElementwiseOp and StableHLO_BinaryElementwiseOp. I see that tests fail without that but why?

Reply:
Yes, this is a surprising point here is the reason: ops inherit everything including `let arguments = ..` or `let extraClassDeclaration = ` from parent `StableHLO_Op`, but if the op itself contains same thing, it will "overwrite" what are inherited.
For example, here `ReduceWindowOp` has 
```
let extraClassDeclaration = [{
    // Get the operation used for reduction applied to `result_index`th result.
    Operation *getReductionOp(int result_index);
   ...
```
I don't think there is a way that we can inherit `let extraClassDeclaration = [{ static bool isCompatibleReturnTypes() ...` from `StableHLO_Op` and somehow **merge** them together (`extraClassDeclaration` is pure string). The related code is https://github.com/llvm/llvm-project/blob/main/mlir/tools/mlir-tblgen/OpDefinitionsGen.cpp#L800 which just read this attribute and use it. Similar situation for StableHLO_UnaryElementwiseOp and StableHLO_BinaryElementwiseOp.

Reply:
Okay, how about a different idea. 

Can we use traits as a mechanism to inherit `isCompatibleReturnTypes` from a base class? E.g. this is how we inherit another static function - `inferReturnType` - in all ops which are marked with `[HLO_CompatibleOperandsAndResultType]`: https://github.com/openxla/stablehlo/blob/b08a70ced3883994b8c410a6439f7114f4325d85/stablehlo/dialect/Base.h#L167-L182
Basically, we could create a new trait which would be added to the list of traits of all StableHLO ops by StableHLO_Op (same for CHLO), and this trait wouldn't do anything except for declaring `isCompatibleReturnTypes`. If individual ops would want to provide their own implementation, they would be able to do that via `extraClassDeclaration`, just like now.

Reply:
Thanks for your suggestion! However, after investigation, there is no clear solution for this approach:
Similar to `HLO_CompatibleOperandsAndResultType`, we can try add a default implementation of `isCompatibleReturnTypes` in `Base.h` or `stablehloOps.cc`, but the errors will be:
1. adding in `Base.h`. There will be implementation conflicts against https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/Interfaces/InferTypeOpInterface.td#L87,
2. adding `Op::isCompatibleReturnTypes` in `stablehloOps.cc` will trigger the error that `Op::isCompatibleReturnTypes` is not declared. And so far there is no clear solution to declare it because of http://google3/third_party/llvm/llvm-project/mlir/tools/mlir-tblgen/OpDefinitionsGen.cpp;l=2269;rcl=475283550, even after we declear it in the trait like:
```
def HLO_CompatibleOperandsAndResultType : TraitList<
[
  DeclareOpInterfaceMethods<InferTypeOpInterface, ["isCompatibleReturnTypes"]>,
  DeclareOpInterfaceMethods<InferShapedTypeOpInterface, ["inferReturnTypeComponents"]>,
  HLO_NativeOpTrait<"CompatibleOperandsAndResultType">
]>;
```
So for now let's stick to the current implementation.